### PR TITLE
Add PermissionManager for Android

### PR DESCRIPTION
**Pull request by cagdasc**
With this issue #32162 I have implemented `PermissionManager` to Godot core. You can request permission/permissions and check it's granted status. This implementation only about with Android and you will get "true" while request permission/s, "false" for check permission from other platforms. I added all existing permissions in Godot export list and mapping with index(`Permission` enum). If everything looks good, I will add documentation. You can find sample project in [repo](https://github.com/cagdasc/godot-samples) and I left a code snippet down.

```
var permission_array: PoolIntArray = []
permission_array.append(PermissionManager.ACCESS_FINE_LOCATION)
permission_array.append(PermissionManager.WRITE_EXTERNAL_STORAGE)
PermissionManager.request_permissions(permission_array)

PermissionManager.request_permission(PermissionManager.CAMERA)

var granted = PermissionManager.check_permission(PermissionManager.CAMERA)
print("CAMERA granted status: ", granted)
```

### CODE REVIEW

**Comment by akien-mga** [comment]
Those permissions are Android-specific, so I don't think it makes sense to bind them as constants for all platforms.

I'm not sure how to best do this though, this type of use case and other similar ones make me want to have platform-specific singletons available in the API, so that we can implement this in an `Android` singleton and not in `OS`.

**Comment by cagdasc** [comment]
@akien-mga thank you for review. For beginning I added all permissions which in  Android exports, I think we can edit this with generic ones. I see `request_permission` function in `OS`. So I add `PermissionsManager` to core. But I can move it to `Android` and `iPhone`. If I make this an `Android` singleton, user have to check is this singleton exist, is this useful?

File: core/os/os.h [hunk 1]
```diff
@@ -530,6 +531,9 @@ class OS {
 	List<String> get_restart_on_exit_arguments() const;
 
 	virtual bool request_permission(const String &p_name) { return true; }
+	virtual bool request_permission_with_index(int p_index) { return true; }
+	virtual bool request_permissions(const PoolIntArray &p_permissions) { return true; };
```
Comment (volzhs) [inline core/os/os.h]: unnecessary `;` at the end?

Comment (cagdasc) [reply core/os/os.h]: Sorry, Fixed up.

**Comment by cagdasc** [comment]
If I do it in this way, I can not implement it to iphone. In different way, I can move `PermissionManager` to `core/os` or wherever do you want and register to platform api like `Javascript` api.

**Comment by cagdasc** [comment]
@akien-mga PR was added __4.0__ milestone. Should I develop it for __vulkan__ branch or will this feature evaluate for __3.2__

**Comment by m4gr3d** [comment]
@cagdasc Unfortunately, this is not at all what I meant in my comment in PR #31598, as this implementation is unnecessarily reinventing the wheels.

For one, you can assume that there's no reasons for a game logic to ask a new permission at runtime. If a permission is needed, it'll be included by the developer at export time by selecting from the list of permissions the export manager makes available. So this means there's no reasons for a permissions API in core.

Given that premise, we already have logic (in the exporter) that inserts the permissions requested by the developer in the final `AndroidManifest.xml` file. All we need to do, is read that manifest file using  [PackageInfo#requestedPermissions](https://developer.android.com/reference/android/content/pm/PackageInfo.html#requestedPermissions) ([reference](https://stackoverflow.com/a/18237962)), automatically request the listed permissions for the app, and broadcast whether or not they were granted.

**Comment by m4gr3d** [comment]
One consideration would be to add a toggle for this logic so that custom build can disable it if they prefer to handle their permissions request process themselves.

**Comment by cagdasc** [comment]
Hi, in Android you request permissions when you make permission sensitive operations and I develop according to this reasons but it can be different in game world. I can convert it like your comment which requesting game start state but I can broadcast it on `OS`. Is that suitable?

**Comment by mathgeniuszach** [comment]
I don't understand why this feature was vetoed. There IS a very good reason to request permissions after install. It's extremely common for apps to put up information walls describing why permissions are asked for _in the app itself_, as opposed to install time, sometimes permissions are optional, and putting those permissions in the install requirements often makes an app seem needlessly suspicious.

As an example, I have custom level support built into my game, and I exported it as an app. Loading custom levels is optional, but in order for the user to be able to easily add levels the folder must be in an accessible location, such as the user's "Documents" folder. To access this, I need at least the "READ_EXTERNAL_STORAGE" permission, though "WRITE_EXTERNAL_STORAGE" is necessary to make the subdirectory for the levels (which means I need "MANAGE_EXTERNAL_STORAGE" for both). The user doesn't need to grant these permissions in order to run the app, but if I ask for it during install time the user just sees a game that's looking to access random files on the sdcard, which looks needlessly suspicious; so instead of denying the permission and using the app anyway, now they have a sour taste and just don't install the app.

So now I'm stuck; either remove custom level support entirely, or include the permission at install time to the detriment of installs for a niche feature.

It is unreasonable to expect that all permissions for an app are given at install time, and even more unreasonable to deny an implementation of this function that was given by a pull request to flesh out a partially implemented function.

**Pull request closed without merging**